### PR TITLE
Handle `emails` field in scim user record

### PR DESCRIPTION
**Pull request by fisx**
https://wearezeta.atlassian.net/browse/WPB-665

## Checklist

 - [x] Add a new entry in an appropriate subdirectory of `changelog.d`
 - [x] Read and follow the [PR guidelines](https://docs.wire.com/developer/developer/pr-guidelines.html)

### CODE REVIEW

File: libs/wire-api/src/Wire/API/User/Scim.hs [hunk 1]
```diff
@@ -353,17 +354,28 @@ data ValidScimId = ValidScimId
...
   arbitrary = do
-    authInfo <- QC.arbitrary
-    -- TODO(fisx): delete qualifiers from name ID
+    authInfo <-
+      QC.arbitrary
+        & fmap (mapThere removeQualifiers)
+        & flip suchThat shortShowNameIDIsNotNothing
     pure $ these onlyThis onlyThat (\_ uref -> onlyThat uref) authInfo
     where
-      -- TODO: the external ID can also be different from the email
```
Comment (fisx) [inline libs/wire-api/src/Wire/API/User/Scim.hs]: i'd leave that in, so we don't forget?

File: libs/wire-api/src/Wire/API/User/Scim.hs [hunk 2]
```diff
@@ -353,17 +354,28 @@ data ValidScimId = ValidScimId
 
 instance Arbitrary ValidScimId where
   arbitrary = do
-    authInfo <- QC.arbitrary
-    -- TODO(fisx): delete qualifiers from name ID
+    authInfo <-
+      QC.arbitrary
+        & fmap (mapThere removeQualifiers)
+        & flip suchThat shortShowNameIDIsNotNothing
```
Comment (fisx) [inline libs/wire-api/src/Wire/API/User/Scim.hs:360]: `unsafeShowNameID` also works with qualifiers, it just adds the extra requirement that qualifiers must not create different overlapping scopes.

also, how can shortShowNameIDIsNotNothing be false on a return value of removeQualifiers?

File: services/spar/src/Spar/Intra/BrigApp.hs [hunk 3]
```diff
@@ -80,6 +80,7 @@ veidFromUserSSOId ::
   m ValidScimId
 veidFromUserSSOId ssoId mEmail = case ssoId of
   UserSSOId uref -> do
+    -- let eid = CI.original $ uref ^. SAML.uidSubject . to SAML.unsafeShowNameID
```
Comment (fisx) [inline services/spar/src/Spar/Intra/BrigApp.hs]: yeah maybe we should do that.

see also: https://wearezeta.atlassian.net/browse/WPB-1506

File: integration/test/Test/Spar.hs [hunk 4]
```diff
@@ -139,8 +142,7 @@ registerUser domain tid email = do
 activateEmail :: (HasCallStack, MakesValue domain) => domain -> String -> App ()
 activateEmail domain email = do
   (key, code) <- bindResponse (BrigInternal.getActivationCode domain email) $ \res -> do
-    res.status `shouldMatchInt` 200
-    key <- res.json %. "key" >>= asString
-    code <- res.json %. "code" >>= asString
-    pure (key, code)
+    (,)
+      <$> (res.json %. "key" >>= asString)
+      <*> (res.json %. "code" >>= asString)
```
Comment (fisx) [inline integration/test/Test/Spar.hs:147]: with this change you're losing the should-match-200 check.

File: services/brig/src/Brig/API/User.hs [hunk 5]
```diff
@@ -306,10 +306,13 @@ createUser new = do
         Just existingAccount ->
           let existingUser = existingAccount.accountUser
               mbSSOid =
-                case (teamInvitation, email, existingUser.userManagedBy) of
+                case (teamInvitation, email, existingUser.userManagedBy, userSSOId existingUser) of
```
[the remark this answers is not in the document]
Comment (battermann) [reply services/brig/src/Brig/API/User.hs:309]: It's a function :)

File: services/spar/test-integration/Test/Spar/Scim/UserSpec.hs [hunk 6]
```diff
@@ -94,6 +94,7 @@ import qualified Wire.API.User.IdentityProvider as User
 import Wire.API.User.RichInfo
 import qualified Wire.API.User.Scim as Spar.Types
 import qualified Wire.API.User.Search as Search
+import qualified Polysemy.Internal.CustomErrors as registered
```
[the remark this answers is not in the document]
Comment (battermann) [reply services/spar/test-integration/Test/Spar/Scim/UserSpec.hs]: Fixed

File: services/spar/test-integration/Test/Spar/Scim/UserSpec.hs [hunk 7]
```diff
@@ -657,7 +658,14 @@ testCreateUserNoIdP = do
...
     liftIO $ userEmail brigUser `shouldBe` Just email
     liftIO $ userManagedBy brigUser `shouldBe` ManagedByScim
-    liftIO $ userSSOId brigUser `shouldBe` Nothing
+    -- TODO(fisx, leif): previous to the change that allows the external ID to be different from the email
+    -- `userSSOId brigUser` was `Nothing`.
+    -- Not sure if it can be a problem that it is now a `Just`?
+    -- The reason is that we now store the external id as the sso_id when the user invitation is created, whereas before
+    -- we stored the email address (as in EmailIdentity) and sso_id was not stored until the user registered.
+    -- Now we need to store the external id when the user is invited because in case it is different from the email, we don't have it
+    -- when the user registers.
```
[the remark this answers is not in the document]
Comment (battermann) [reply services/spar/test-integration/Test/Spar/Scim/UserSpec.hs:668]: Ok, to me the way it is now, seems correct, anyway.

File: services/spar/test-integration/Test/Spar/Scim/UserSpec.hs [hunk 8]
```diff
@@ -1733,17 +1742,6 @@ testUpdateExternalId withidp = do
 
   do
     -- email changes to other email
-    email <- randomEmail
```
[the remark this answers is not in the document]
Comment (battermann) [reply services/spar/test-integration/Test/Spar/Scim/UserSpec.hs]: It's just a refactoring making it more DRY. It just looks weird in the diff.

File: integration/test/Test/Spar.hs [hunk 9]
```diff
@@ -90,6 +92,38 @@ testSparExternalIdDifferentFromEmail = do
...
+    updatedScimUser <- setField "emails" (Array (Vector.fromList [object ["value" .= newEmail]])) scimUserWith2Updates
+    currentExtId <- updatedScimUser %. "externalId" >>= asString
+    bindResponse (updateScimUser OwnDomain tok userId updatedScimUser) $ \res -> do
+      res.status `shouldMatchInt` 200
+
+    bindResponse (findUsersByExternalId OwnDomain tok currentExtId) $ \res -> do
+      res.status `shouldMatchInt` 200
+      u <- res.json %. "Resources" >>= asList >>= assertOne
+      u %. "externalId" `shouldMatch` currentExtId
+      (u %. "emails" >>= asList >>= assertOne >>= (%. "value")) `shouldMatch` oldEmail
```
[the remark this answers is not in the document]
Comment (battermann) [reply integration/test/Test/Spar.hs:111]: This is valid an I fixed it.

File: services/spar/src/Spar/Scim/User.hs [hunk 10]
```diff
@@ -979,6 +979,7 @@ synthesizeStoredUser acc veid =
       let (createdAt, lastUpdatedAt) = fromMaybe (now, now) accessTimes
 
       handle <- lift $ Brig.giveDefaultHandle acc.account.accountUser
+      -- TODO(fisx): do we need to consider unvalidated emails here as well?
```
[the remark this answers is not in the document]
Comment (battermann) [reply services/spar/src/Spar/Scim/User.hs:982]: Fixed

File: integration/test/Test/Spar.hs (hunk 4 above)
Comment (battermann) [reply integration/test/Test/Spar.hs:147]: The 200 check does not provide much value, IMO.

File: integration/test/Test/Spar.hs [hunk 11]
```diff
@@ -26,3 +34,266 @@ testSparUserCreationInvitationTimeout = do
...
+    (u %. "emails" >>= asList >>= assertOne >>= (%. "value")) `shouldMatch` email
+  bindResponse (getUsersId OwnDomain [userId]) $ \res -> do
+    res.status `shouldMatchInt` 200
+    u <- res.json >>= asList >>= assertOne
+    u %. "email" `shouldMatch` email
+    subject <- u %. "sso_id.subject" >>= asString
+    subject `shouldContainString` extId
+    u %. "handle" `shouldMatch` (scimUser %. "userName")
+
+  -- Verify that updating the scim user works
```
Comment (fisx) [inline integration/test/Test/Spar.hs:60]:
```suggestion
  -- Verify that updating `userName` (handle) works
```

File: integration/test/Test/Spar.hs [hunk 12]
```diff
@@ -26,3 +34,266 @@ testSparUserCreationInvitationTimeout = do
...
+
+testSparExternalIdDifferentFromEmail :: (HasCallStack) => App ()
+testSparExternalIdDifferentFromEmail = do
+  (owner, tid, _) <- createTeam OwnDomain 1
+  tok <- createScimToken owner >>= \resp -> resp.json %. "token" >>= asString
+  email <- randomEmail
+  extId <- randomExternalId
+  scimUser <- randomScimUserWith extId email
+  userId <- createScimUser OwnDomain tok scimUser >>= getJSON 201 >>= (%. "id") >>= asString
+  registerUser OwnDomain tid email
```
Comment (fisx) [inline integration/test/Test/Spar.hs:142]: inspect situation *before* registration? maybe just mv L143-151 into a helper function and all that twice?

File: services/spar/src/Spar/Intra/BrigApp.hs [hunk 13]
```diff
@@ -64,58 +65,69 @@ import Spar.Sem.GalleyAccess (GalleyAccess)
...
 -- settings and is now onboarded to saml/scim.  If this case can safely be ruled out, it's ok
 -- to just set it to 'Nothing'.
-veidFromBrigUser :: (MonadError String m) => User -> Maybe SAML.Issuer -> m ValidExternalId
-veidFromBrigUser usr mIssuer = case (userSSOId usr, userEmail usr, mIssuer) of
-  (Just ssoid, _, _) -> veidFromUserSSOId ssoid
-  (Nothing, Just email, Just issuer) -> pure $ EmailAndUref email (SAML.UserRef issuer (fromRight' $ emailToSAMLNameID email))
-  (Nothing, Just email, Nothing) -> pure $ EmailOnly email
+--
+-- `userSSOId usr` can be empty if the user has no SAML credentials and is brought under scim
+-- management for the first time with `getUserById`.  In that case, the externalId is taken to
```
Comment (fisx) [inline services/spar/src/Spar/Intra/BrigApp.hs:104]:
```suggestion
-- management for the first time.  In that case, the externalId is taken to
```

File: libs/wire-api/src/Wire/API/User/Scim.hs [hunk 14]
```diff
@@ -339,51 +341,64 @@ data ValidScimUser = ValidScimUser
   }
   deriving (Eq, Show)
 
--- | Note that a 'SAML.UserRef' may contain an email. Even though it is possible to construct a 'ValidExternalId' from such a 'UserRef' with 'UrefOnly',
--- this does not represent a valid 'ValidExternalId'. So in case of a 'UrefOnly', we can assume that the 'UserRef' does not contain an email.
-data ValidExternalId
-  = EmailAndUref EmailAddress SAML.UserRef
-  | UrefOnly SAML.UserRef
-  | EmailOnly EmailAddress
+-- | This type carries parsed externalId, email address, and saml credentials. It contains the email address and saml credentials,
```
Comment (fisx) [inline libs/wire-api/src/Wire/API/User/Scim.hs:344]:
```suggestion
-- | This type carries externalId, plus email address (validated if present, unvalidated if not) and saml credentials,
```

File: libs/wire-api/src/Wire/API/User/Scim.hs [hunk 15]
```diff
@@ -339,51 +341,64 @@ data ValidScimUser = ValidScimUser
...
+-- because those are sometimes derived from the externalId field.
+--
+-- Note that a 'SAML.UserRef' may contain an email. Even though it is possible to construct a 'ValidScimId' from such a 'UserRef' with 'UrefOnly',
+-- this does not represent a valid 'ValidScimId'. So in case of a 'UrefOnly', we can assume that the 'UserRef' does not contain an email.
```
Comment (fisx) [inline libs/wire-api/src/Wire/API/User/Scim.hs:348]:
```suggestion
```

File: services/brig/src/Brig/Data/User.hs [hunk 16]
```diff
@@ -160,7 +161,8 @@ newAccountInviteViaScim uid tid locale name email = do
     user domain loc =
       User
         (Qualified uid domain)
-        (Just $ EmailIdentity email)
+        -- TODO: do we really want to store the email of an unvalidated/pending account invitation?
```
Comment (fisx) [inline services/brig/src/Brig/Data/User.hs:164]:
```suggestion
```

(the answer is "yes".)

File: services/brig/src/Brig/Team/API.hs [hunk 17]
```diff
@@ -166,7 +166,7 @@ createInvitationViaScim ::
   TeamId ->
   NewUserScimInvitation ->
   (Handler r) UserAccount
-createInvitationViaScim tid newUser@(NewUserScimInvitation _tid uid loc name email role) = do
+createInvitationViaScim tid newUser@(NewUserScimInvitation _tid uid _eid loc name email role) = do
```
[the remark this answers is not in the document]
Comment (fisx) [reply services/brig/src/Brig/Team/API.hs:169]: ah, it binds `newUser`, which is passed on to `createUserInviteViaScim`, which in turn wants `eid`.  all good!

**Review by battermann — APPROVED** [review]

**Pull request merged**